File: frontend/lib/reports.ts
import { apiFetch } from "@/lib/api-client";

export type Overview = {
  total_revenue: number;
  new_deals: number;
  revenue_at_risk: number;
  revenue_at_risk_customer_count: number;
  open_invoices: number;
  deltas: Record<string, number>;
};

export type RevenueTrendPoint = {
  day_of_period: number;
  this_period: number;
  last_period: number | null;
};

export type RevenueBySourceSlice = { label: string; revenue: number; pct: number };

export type TopCustomer = {
  id: string;
  name: string;
  revenue_total: number;
  deal_count: number;
  relationship_score: number | null;
};

export function getOverview(token: string, period = "this_month"): Promise<Overview> {
  return apiFetch<Overview>(`/reports/overview?period=${period}`, { token });
}

export function getRevenueTrend(token: string): Promise<RevenueTrendPoint[]> {
  return apiFetch<RevenueTrendPoint[]>("/reports/revenue-trend", { token });
}

export function getRevenueBySource(token: string): Promise<RevenueBySourceSlice[]> {
  return apiFetch<RevenueBySourceSlice[]>("/reports/revenue-by-source", { token });
}

export function getTopCustomers(token: string): Promise<TopCustomer[]> {
  return apiFetch<TopCustomer[]>("/reports/top-customers", { token });
}

export type ScoreHistoryPoint = {
  score: number;
  algorithm_version: string;
  computed_at: string;
};

export function getOrganizationScoreHistory(token: string): Promise<ScoreHistoryPoint[]> {
  return apiFetch<ScoreHistoryPoint[]>("/reports/score-history?entity_type=organization", { token });
}

File: backend/app/schemas/reports.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class OverviewOut(BaseModel):
    total_revenue: float
    new_deals: int
    revenue_at_risk: float
    revenue_at_risk_customer_count: int
    open_invoices: int
    deltas: dict[str, float]


class RevenueTrendPoint(BaseModel):
    day_of_period: int
    this_period: float
    last_period: float | None


class RevenueBySourceSlice(BaseModel):
    label: str
    revenue: float
    pct: float


class TopCustomerOut(BaseModel):
    id: uuid.UUID
    name: str
    revenue_total: float
    deal_count: int
    relationship_score: int | None


class ScoreHistoryPoint(BaseModel):
    score: int
    algorithm_version: str
    computed_at: datetime
